### PR TITLE
Lint on some incorrect uses of mem::zeroed / mem::uninitialized

### PR DESCRIPTION
**Pull request by RalfJung**
Cc https://github.com/rust-lang/rust/issues/62825 and https://internals.rust-lang.org/t/make-mem-uninitialized-and-mem-zeroed-panic-for-some-types-where-0-is-a-niche/10605

This does not yet handle `NonNull`/`NonZero*`, but it is a start.

I also improved some doc issues I hit on the way, and added a useful helper to `TyS`.

EDIT: I added the relnotes label mostly as a proposal -- I think this is worth mentioning, but leave the decision up to the release team.

### CODE REVIEW

File: src/librustc_lint/builtin.rs [hunk 1]
```diff
@@ -1862,3 +1862,90 @@ impl EarlyLintPass for IncompleteFeatures {
             });
     }
 }
+
+declare_lint! {
+    pub INVALID_VALUE,
```
Comment (RalfJung) [inline src/librustc_lint/builtin.rs]: I used the "validity" terminology here. Does that seem okay?

File: src/librustc/ty/mod.rs [hunk 2]
```diff
@@ -1843,6 +1843,7 @@ pub struct VariantDef {
     /// Flags of the variant (e.g. is field list non-exhaustive)?
     flags: VariantFlags,
     /// Recovered?
+    // FIXME: Needs proper doc. Recovered whom from what?
```
Comment (petrochenkov) [inline src/librustc/ty/mod.rs]: From some syntactic error.
So e.g. the field list may be incomplete, and we may want to suppress some errors related to that.

Comment (petrochenkov) [reply src/librustc/ty/mod.rs]: Introduced in https://github.com/rust-lang/rust/pull/59266.
The recovery terminology is so ubiquitous in the frontend that we assumed it's obvious.

Comment (RalfJung) [reply src/librustc/ty/mod.rs]: So what would be an appropriate doc comment?
```
/// Variant is obtained as part of recovering from a syntactic error.
/// May be incomplete or bogus.
```

File: src/librustc_lint/builtin.rs [hunk 3]
```diff
@@ -1862,3 +1862,90 @@ impl EarlyLintPass for IncompleteFeatures {
...
+    Warn,
+    "an invalid value is being created (such as a NULL reference)"
+}
+
+declare_lint_pass!(InvalidValue => [INVALID_VALUE]);
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for InvalidValue {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &hir::Expr) {
+
+        const ZEROED_PATH: &[Symbol] = &[sym::core, sym::mem, sym::zeroed];
```
Comment (bjorn3) [inline src/librustc_lint/builtin.rs]: Does this work for std too?

Comment (RalfJung) [reply src/librustc_lint/builtin.rs]: Yes, it follows reexports. See the test case I added, which uses names from `std`.

File: src/librustc/ty/mod.rs (hunk 2 above)
Comment (RalfJung) [reply src/librustc/ty/mod.rs]: > The recovery terminology is so ubiquitous in the frontend that we assumed it's obvious.

`Ty` is shared with the middle-end though, and as someone who did a lot of work on the MIR level I have no idea about recovery. :)

**Comment by RalfJung** [comment]
>  @bjorn3 deleted a comment

Please don't do that. GH has a "hide comment" feature and "resolve discussion" which helps not to clutter the main thread. But don't delete comments after other people replied, and really don't delete other people's comments.

**Comment by bjorn3** [comment]
> But don't delete comments after other people replied, and really don't delete other people's comments.

I though I removed it before someone commented. I think your comment got deleted because I deleted mine.

> GH has a "hide comment" feature

Will use it in the future.

File: src/librustc_lint/builtin.rs [hunk 4]
```diff
@@ -1862,3 +1862,92 @@ impl EarlyLintPass for IncompleteFeatures {
...
+                }
+                Tuple(..) => {
+                    // Proceed recursively, check all fields.
+                    ty.tuple_fields().all(|field| ty_maybe_allows_zero_init(tcx, field))
+                }
+                // FIXME: Would be nice to also warn for `NonNull`/`NonZero*`.
+                // FIXME: *Only for `mem::uninitialized`*, we could also warn for `bool`,
+                //        `char`, and any multivariant enum.
+                // Conservative fallback.
+                _ => true,
```
Comment (jethrogb) [inline src/librustc_lint/builtin.rs]: Can you check for types with `rustc_layout_scalar_valid_range_start` here?

Comment (RalfJung) [reply src/librustc_lint/builtin.rs]: In the future this should probably check for that attribute, yes.

**Comment by RalfJung** [comment]
Would be nice if this could land before beta cutoff (Tuesday), so that we get these lints on stable.

**Review by eddyb — APPROVED** [review]

**Comment by RalfJung** [comment]
Looks like this [breaks clippy tests](https://github.com/rust-lang/rust/pull/63412#issuecomment-520024474).

**Comment by RalfJung** [comment]
Blocked on clippy PR: https://github.com/rust-lang/rust-clippy/pull/4368

**Comment by RalfJung** [comment]
@bors p=1

Would be good to get that lint in together with https://github.com/rust-lang/rust/pull/63343

**Pull request merged**